### PR TITLE
Resolve crasher 28704

### PR DESCRIPTION
**Pull request by hartbit**
This crash happens because shouldAddSelfFixit (a check for the expected_self_before_reference diagnotic) performs name lookup inside the parser, which causes a search for extensions, which triggers the mangler, which can get confused if we are in an unfinished declaration (a PatternBindingInitilizer in this case).

A cleaner fix would be to move that logic inside Sema, but it would require much more refactoring that I don’t feel comfortable doing. Instead, this band-aid checks if the innermost type is inside a local context. If that’s the case, we can ignore the search for extensions.

### CODE REVIEW

**Review by nkcsgexi — COMMENTED** [review]
This crash seems to me very internal to ASTMangler. Could we just add a null check early return in ASTMangler.cpp:1047? @eeckstein

**Comment by slavapestov** [comment]
Why does name lookup call into the mangler at all? I'd like to identify the root cause of this issue and fix it there instead of adding workarounds either in the parser or the mangler.

**Comment by CodaFi** [comment]
Parse shouldn't be doing Lookup at all is the actual problem.

**Comment by slavapestov** [comment]
Right, but why does the mangler get called during type checking? We shouldn't have to mangle anything until SILGen.

**Comment by hartbit** [comment]
From what I understand, name lookup loads the type's extensions (to be able to lookup their members), which in turn calls into `ModuleFile::loadExtensions`, which mangles the type to compare it to the extensions it found (I guess to make sure they are the right extensions).

File: lib/Parse/ParseExpr.cpp [hunk 1]
```diff
@@ -1869,12 +1869,18 @@ DeclName Parser::parseUnqualifiedDeclName(bool afterDot,
...
+
   if (auto *Nominal = Current->getInnermostTypeContext()->
       getAsNominalTypeOrNominalTypeExtensionContext()){
     // FIXME: we cannot resolve members appear later in the body of the nominal.
-    auto LookupResults = Nominal->lookupDirect(Name);
+    
+    // Fixes a crasher where a lookup for a type in a local context would
+    // crash inside ASTMangler
+    auto IgnoreExtensions = InnermostTypeContext->getLocalContext() != nullptr;
+    auto LookupResults = Nominal->lookupDirect(Name, IgnoreExtensions);
```
Comment (jrose-apple) [inline lib/Parse/ParseExpr.cpp]: If we have to do something here, let's just teach `lookupDirect` and such that declarations in local contexts can't have extensions.

Comment (hartbit) [inline lib/Parse/ParseExpr.cpp]: So move the `InnermostTypeContext->getLocalContext() != nullptr` check inside lookupDirect?

Comment (jrose-apple) [inline lib/Parse/ParseExpr.cpp]: Probably further down, to whatever decides whether we need to go look for loaded extensions, but yes. (Don't forget a test case with a type inside a type inside a function.)

**Comment by slavapestov** [comment]
Oh, I like @jrose-apple's suggestion the best.

Extensions for local types make no sense -- the extension would have to be defined inside the same scope as the type, and at that point you may as well put the extension's members inside the type itself.

File: lib/Parse/ParseExpr.cpp (hunk 1 above)
Comment (slavapestov) [inline lib/Parse/ParseExpr.cpp]: I think NominalTypeDecl::prepareExtensions() might be a good place to look.

Comment (hartbit) [inline lib/Parse/ParseExpr.cpp]: Done!

**Comment by hartbit** [comment]
@slavapestov Let me know what you think of the current fix

**Comment by slavapestov** [comment]
Looks good. Eventually we might want to allow extensions in local context to allow local types to conditionally conform to protocols, since we have no other way of writing that, but in any case the "find stuff to load" code path won't be needed.

**Comment by slavapestov** [comment]
@swift-ci Please smoke test

**Comment by jrose-apple** [comment]
By the way, in the future can you put more descriptive titles in your PRs? That helps us figure out which PRs we'd be interested in / able to help with.

**Comment by hartbit** [comment]
@jrose-apple What would you suggest for a PR that fixes a crash like this? The title of the crasher? Something else?

**Comment by jrose-apple** [comment]
Most of the crashers were generated by a fuzzer; it's not super interesting. I think just describe the scenario: "Fix a crash where lookup within a local type tries to look for extensions".

**Comment by jrose-apple** [comment]
I sympathize that for this particular bug the crash location, the cause, and the fix are all particularly far apart and that makes it hard to choose a title!

**Comment by slavapestov** [comment]
The Linux failure seems unrelated. Let's try again

**Comment by slavapestov** [comment]
@swift-ci Please smoke test Linux

**Comment by hartbit** [comment]
@slavapestov I think we may want to wait before merging. Recent crasher 28721 is basically the same issue and can be solved with a small change.

**Comment by hartbit** [comment]
@slavapestov this solves both crashers. hope it passes tests.

**Comment by slavapestov** [comment]
@swift-ci Please smoke test

**Comment by slavapestov** [comment]
If the tests pass, I will squash this and merge. In the future please rebase and force push to your development branch instead of pushing merge commits. It helps keep our git history tidy. :-)

**Comment by slavapestov** [comment]
@hartbit 5 more patches and you will make the top 100 contributors

**Pull request merged**